### PR TITLE
FPGA: Update invocation interfaces tutorial

### PR DESCRIPTION
**Pull request by whitepau**
# Existing Sample Changes
## Description

Update invocation interfaces code sample to demonstrate streaming/register-mapped invocation interfaces, as well as register-mapped and conduit kernel arguments. Also demonstrate some corner cases:

* how to cast away annotated_arg to access C++ members
* how to use pipelined streaming invocation interface
* how to disable read_in on a pipelined streaming invocation interface

Remove `piplined_kernels` code sample as this PR makes it obsolete.

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Implement fixes for ONSAM Jiras

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Command Line
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [X] VSCode
- [ ] When compiling the compiler flag "-Wall -Wformat-security -Werror=format-security" was used
- [X] Windows FPGA regtest: https://spetc.intel.com/testanalysis?trview=0&testRunIds=8778883&tapgsize=20
- [X] Linux FPGA regtest: https://spetc.intel.com/testanalysis?trview=0&reporting_eng=true&runBy_reporting_eng=true&testRunIds=8778879&tapgsize=20

### CODE REVIEW

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/pipelined_kernels/sample.json [hunk 1]
```diff
@@ -1,71 +0,0 @@
-{
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/pipelined_kernels/sample.json:1]: If removing the pipelined_kernels sample, it needs to be removed from the top level readme https://github.com/oneapi-src/oneAPI-samples/blob/development/DirectProgramming/C%2B%2BSYCL_FPGA/README.md

**Review by yuguen-intel — CHANGES_REQUESTED** [review]

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/README.md [hunk 2]
```diff
@@ -307,55 +426,67 @@ void TestLambdaStreamingKernel(sycl::queue &q, ValueT *in, ValueT *out, size_t c
...
+   ./reg_map_lambda.fpga_emu
+   ./stream_lambda.fpga_emu
    ```
 2. Run the sample on the FPGA simulator.
    ```
-   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./register_map_functor_model.fpga_sim
-   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./streaming_functor_model.fpga_sim
-   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./register_map_lambda_model.fpga_sim
-   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./streaming_lambda_model.fpga_sim
+   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./reg_map_functor.fpga_sim
```
[the remark this answers is not in the document]
Comment (whitepau) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/README.md:456]: agree

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/README.md [hunk 3]
```diff
@@ -56,155 +56,274 @@ You can also find more information about [troubleshooting build errors](/DirectP
...
+By default, your IP's `start` and `done` signals will appear in the IP's CSR. This is true whether you declare your kernel using the 'functor' or 'lambda' syntax.
+
+[//]: # "Comment: Use an HTML Table to render code snippets side-by-side" 
+
+<table> 
+<tr>
+<th>Functor Syntax</th> <th>Lambda Syntax</th>
+</tr>
+<tr>
+<td>
```
[the remark this answers is not in the document]
Comment (whitepau) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/README.md:105]: removed

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/pipelined_kernels/sample.json (hunk 1 above)
Comment (whitepau) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/pipelined_kernels/sample.json:1]: removed. i noticed that invocation_interfaces was missing as well, so i added it

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/src/reg_map_functor.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    if (!d.has(sycl::aspect::usm_host_allocations)) {
+      std::cerr << "ERROR: The selected device does not support USM host"
+                << " allocations" << std::endl;
+      return 1;
+    }
+
+    int *input = sycl::malloc_host<int>(count, q);
+    int *functor_register_map_out = sycl::malloc_host<int>(count, q);
+    int *golden_out = sycl::malloc_host<int>(count, q);
+
```
[the remark this answers is not in the document]
Comment (whitepau) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/src/reg_map_functor.cpp:87]: fixed

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/src/stream_functor.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                << " allocations" << std::endl;
+      return 1;
+    }
+
+    Point input;
+    input.x = 1;
+    input.y = 'a';
+
+    Point *functor_streaming_out = sycl::malloc_host<Point>(count, q);
+    Point *golden_out = sycl::malloc_host<Point>(count, q);
```
[the remark this answers is not in the document]
Comment (whitepau) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/src/stream_functor.cpp:104]: fixed

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/src/stream_pipelined.cpp [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    if (!d.has(sycl::aspect::usm_host_allocations)) {
+      std::cerr << "ERROR: The selected device does not support USM host"
+                << " allocations" << std::endl;
+      return 1;
+    }
+
+    int *input = sycl::malloc_host<int>(count, q);
+    int *functor_streaming_pipelined_out = sycl::malloc_host<int>(count, q);
+    int *golden_out = sycl::malloc_host<int>(count, q);
```
[the remark this answers is not in the document]
Comment (whitepau) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/src/stream_pipelined.cpp:76]: fixed

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/src/stream_rm_stall.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    if (!d.has(sycl::aspect::usm_host_allocations)) {
+      std::cerr << "ERROR: The selected device does not support USM host"
+                << " allocations" << std::endl;
+      return 1;
+    }
+
+    int *input = sycl::malloc_host<int>(count, q);
+    int *functor_streaming_rm_downstream_stall_out =
+        sycl::malloc_host<int>(count, q);
+    int *golden_out = sycl::malloc_host<int>(count, q);
```
[the remark this answers is not in the document]
Comment (whitepau) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/invocation_interfaces/src/stream_rm_stall.cpp:90]: fixed

**Review by yuguen-intel — APPROVED** [review]

**Pull request merged**